### PR TITLE
Feature/20169

### LINKED ISSUE #20169: Allow for JUnitPlatformTestClassProcessor filtering to support custom test sources
Since JUnit 5 allows for custom TestSources to be configured, JUnitPlatformTestClassProcessor could provide a way for those sources to be recognized and filterable. This way, custom method sources will enjoy the same filtering capabilities that method sources do. 

For now, this change specifically targets ArchUnit's `FieldSource`, and would like to use the solution provided to solve [the following issue](https://github.com/TNG/ArchUnit/issues/641). 

### Expected Behavior
It is possible to define a way in which class name and test name can be extracted for filtering custom test sources in `JUnitPlatformTestClassProcessor`

### Current Behavior
`JUnitPlatformTestClassProcessor` is currently hardcoded to only recognize `MethodSource`s

### Context
Please have a look on the attached draft pull request. I have verified that the PR solves the issue for ArchUnit using Java SPI, but I feel it's not a very Gradle-friendly solution (note the need for mingling the classloader of `test-junit-platform` and test classes needed to load both out-of-the-box and custom `TestSelectorFactory` implementations). 

I have no experience extending Gradle and thus, I'm not sure what a proper extension should look like. I'd be really grateful  if you could point me in the right direction. I'd prefer to keep the solution drop-in or requiring the application of a Gradle plugin at worst, if at all possible. 

(I could see for example that `JUnitPlatformTestClassProcessorFactory` locates some of the dependencies using a `ServiceRegistry`, so I thought maybe I could somehow leverage that? Though I'm not sure how to put stuff in the registry from inside a custom Gradle plugin. Is this even the right direction?) 

If you need more context, please have a look at a [Surefire issue](https://issues.apache.org/jira/browse/SUREFIRE-2037) which aims to solve the same issue for Maven.

### PR DESCRIPTION
**Pull request by crizzis**
Fixes #20169

### Context
Support for custom JUnit 5 `TestSource`s in `JUnitPlatformTestClassProcessor`. Please see related [GtHub issue](https://github.com/gradle/gradle/issues/20169)

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/tasks/TaskExecutionException.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
  */
 @Contextual
 public class TaskExecutionException extends DefaultMultiCauseException {
-    private final Task task;
+    private transient final Task task;
```
Comment (crizzis) [inline subprojects/core/src/main/java/org/gradle/api/tasks/TaskExecutionException.java:27]: I noticed when running (Gradle via the Tooling API) that sometimes, an attempt is made to serialize this field, hence the `transient` modifier

**Comment by crizzis** [comment]
> This pull request has been automatically marked as stale because it has not had recent activity. Given the limited bandwidth of the team, it will be closed if no further activity occurs. If you intend to work on this pull request, please ask the team to reopen the PR or push a new PR. Thank you for your contributions.

@jbartok @big-guy FYI This is still relevant

**Comment by jbartok** [comment]
@bot-gradle test QFL

**Comment by jbartok** [comment]
Thank you for the effort. Unfortunately, this is not something we can incorporate right away (due to a lack of resources to review and understand). We'll keep it in our queue.

**Comment by ov7a** [comment]
@gradle/bt-jvm should the contributor do the updates to the PR or given the state of this PR it should be closed?

**Comment by big-guy** [comment]
We think we should fix the underlying issue: https://github.com/gradle/gradle/issues/20169#issuecomment-1954698406

When we get back around to looking at this, we may be able to resurrect part of this PR. I'm going to close it for now.

**Pull request closed without merging**